### PR TITLE
Karinna Iniguez - API Muncher - Octos

### PR DESCRIPTION
**Pull request by karinnainiguez**
# API Muncher
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
How did you go about exploring the Edamam API, how did you try querying the API? |  I read through the documentation in Edamam and once I had an idea of the requirements in a query, I tested a few different things in Postman. 
Describe your API Wrapper.  How did you decide on the methods you created? | The Edamam api has two different ways of searching.  Either with a query, or with the recipe id.  So, I made a method for searching with both of those methods.  My index controller action uses the method .search(query) to search based on a word or phrase.  My show controller action uses the .singlesearch(id) method to search for one specific recipe. 
Describe an edge case or failure case test you wrote for your API Wrapper. | I wrote a test for the case that the search does not return any "hits".  In that case, I made the business decision to have it return nil instead of a search.
Explain how VCR aids in testing an API. |  When using an API, we do not want to rely on the API being available for testing.  A way to reduce the number of requests sent to the API, and allow for offline testing, is to record previous responses from the API.  VCR does that by recording the response it gets from the API in a 'cassette'.  When the tests are run again, if the search query is the same as it previously was, it will not have to go out to the API< it can use the recorded version of the response.
What is the Heroku URL of your deployed application? |  https://yummmmster.herokuapp.com/
Provide a link to the Trello board you used | https://trello.com/b/7MnmzK3D/api-muncher

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,8 @@
+Rails.application.routes.draw do
+  # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+
+  root 'recipes#search'
+  get 'recipes', to: 'recipes#index', as: 'results'
+  get 'recipe/:id', to: 'recipes#show', as: 'recipe'
```
Comment (droberts-sea) [inline config/routes.rb:6]: Could you use `resources` for the index and show pages?

File: app/views/recipes/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,25 @@
+<header>
+  <%= render partial: 'search_form' %>
+</header>
```
Comment (droberts-sea) [inline app/views/recipes/index.html.erb:3]: Since you include the search form on every page, it might make sense to put it in `application.html.erb`.

File: app/views/recipes/_search_form.html.erb [hunk 3]
```diff
@@ -0,0 +1,11 @@
+
+<section class="search-form">
+  <h1><%= link_to "YUMMMMSTER", root_path %></h1>
+  <%= form_tag results_path, method: :get do %>
```
Comment (droberts-sea) [inline app/views/recipes/_search_form.html.erb:4]: Good use of a view partial to DRY this up.

File: lib/edamam_api_wrapper.rb [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  def self.single_search(id)
+    url = URI.encode("#{BASE_URL}/search?r=#{REC_ID_URL}#{id}&app_id=#{EDAMID}&app_key=#{KEY}&to=100")
+
+    response = HTTParty.get(url)
+
+    if response[0]
+      recipe = Recipe.new(response[0])
+      return recipe
+    else
+
```
Comment (droberts-sea) [inline lib/edamam_api_wrapper.rb:34]: You should probably have something in this `else` clause - return `nil` or throw an exception.

File: app/controllers/recipes_controller.rb [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+    @query = params[:query]
+    search = EdamamApiWrapper.search(@query)
+    if search
+      @recipes = search.recipes.paginate(page: params[:page], per_page: 12)
+    end
+  end
+
+  def show
+    @recipe = EdamamApiWrapper.single_search(params[:id])
+  end
```
Comment (droberts-sea) [inline app/controllers/recipes_controller.rb:15]: What if the search fails? Your page should probably report the failure to the user.

File: test/controllers/recipes_controller_test.rb [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        get results_path, params: {query: query}
+      end
+
+      must_respond_with :success
+    end
+
+    it "responds with success for search with no results" do
+      VCR.use_cassette("recipes") do
+        query = "something that doesnt exist"
+        get results_path, params: {query: query}
```
Comment (droberts-sea) [inline test/controllers/recipes_controller_test.rb:25]: Good job catching this error case!

File: test/controllers/recipes_controller_test.rb [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+      end
+
+      must_respond_with :success
+    end
+  end
+
+  describe "show" do
+    it "responds with success" do
+      VCR.use_cassette("recipes") do
+        id = "67b2867627c159e4f5bbaf2431b3ccb6"
```
Comment (droberts-sea) [inline test/controllers/recipes_controller_test.rb:35]: What if you give it a bogus URI?

File: test/controllers/recipes_controller_test.rb (hunk 6 above)
Comment (droberts-sea) [reply test/controllers/recipes_controller_test.rb:25]: It might also be worthwhile to add some tests around the paging parameters:

- What happens if they're absent?
- Do you get different results when they change?
- What if you send a bogus value, like a negative page number?

File: test/lib/edamam_api_wrapper_test.rb [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+      query = "something that doesnt exist"
+      response = EdamamApiWrapper.search(query)
+      response.must_be_nil
+    end
+  end
+
+  it "returns single recipe if searching single" do
+    VCR.use_cassette("recipes") do
+      id = "67b2867627c159e4f5bbaf2431b3ccb6"
+      response = EdamamApiWrapper.single_search(id)
```
Comment (droberts-sea) [inline test/lib/edamam_api_wrapper_test.rb:23]: What happens for an invalid ID?

**Comment by droberts-sea** [comment]
# API Muncher
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Core Requirements** | 
Git hygiene | yes
Comprehension questions | yes
**General** | 
Rails fundamentals (RESTful routing, use of named paths) | yes
Semantic HTML | yes
Errors are reported to the user | no - when I mangle the recipe URI, I get an exception rather than a polite error message
API Wrapper to handle the API requests | yes
Controller testing | yes
Lib testing | yes
Search Functionality | yes
List Functionality | yes
Show individual item functionality (link to original recipe opens in new tab) | yes (doesn't pop out into new tab)
**Styling** | 
Responsive layout | yes
List View shows 10 items at a time/pagination | yes
The app is styled to create an attractive user interface | yes - looks great!
**API Features** | 
The App attributes Edaman | yes
The VCR casettes do not contain the API key | yes
**External Resources** | 
Link to Trello Board | yes
Link to deployed app on Heroku | yes
**Overall** | Great work overall! Your code is well-organized and easy to read, and it's clear the core learning goals of the assignment were met.<br><br>The one piece that's missing is error handling for the recipe show workflow - from top to bottom your app lacks a story on how to figure out the API call didn't work and report the error to the user. This impacts your views, controller, lib files and tests.<br><br>Other than that, I'm quite happy with this submission. Keep up the hard work!
